core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs

```diff
@@ -49,30 +48,178 @@ public override void addPassword(Scheme scheme, int port, string hostName, strin
+                CRED_TYPE_GENERIC, 0, CRED_PERSIST_ENTERPRISE);
+
+            if (protocol.isTokenConfigurable())
+            {
+                winCred.Attributes["Token"] = credential.getToken();
+            }
+            if (protocol.isOAuthConfigurable())
+            {
+                winCred.Attributes["OAuth Access Token"] = credential.getOauth().getAccessToken();
+                winCred.Attributes["OAuth Refresh Token"] = credential.getOauth().getRefreshToken();
+                if (credential.getOauth().getExpiryInMilliseconds() != null)
+                {
+                    winCred.Attributes["OAuth Expiry"] = credential.getOauth().getExpiryInMilliseconds().longValue().ToString();
+                }
+            }
+            if (protocol.isPrivateKeyConfigurable())
+            {
+                winCred.Attributes["Private Key Passphrase"] = credential.getIdentityPassphrase();
+            }
+            if (credential.isPasswordAuthentication())
+            {
+                if (string.IsNullOrWhiteSpace(credential.getUsername()))
+                {
+                    logger.warn(string.Format("No username in credentials for bookmark {0}", bookmark.getHostname()));
+                    return;
+                }
+                if (string.IsNullOrWhiteSpace(credential.getPassword()))
+                {
+                    logger.warn(string.Format("No password in credentials for bookmark {0}", bookmark.getHostname()));
+                    return;
+                }
+            }
+            if (!WinCredentialManager.SaveCredentials(target.AbsoluteUri, winCred))
+            {
+                base.save(bookmark);
+            }
         }
 
-        public override void deletePassword(Scheme scheme, int port, string hostName, string user)
+        private static Uri ToUri(Host bookmark)
         {
-            var hostUrl = hostUrlProvider.get(scheme, port, user, hostName, string.Empty);
-            WinCredentialManager.RemoveCredentials(hostUrl);
+            var protocol = bookmark.getProtocol();
+            var credentials = bookmark.getCredentials();
+
+            var targetBuilder = new UriBuilder(PreferencesFactory.get().getProperty("application.container.name"), string.Empty);
+            var pathBuilder = new StringBuilder();
+            pathBuilder.Append(protocol.getIdentifier());
+            if (protocol.isHostnameConfigurable() || !(protocol.isTokenConfigurable() || protocol.isOAuthConfigurable()))
+            {
+                pathBuilder.Append(":" + bookmark.getHostname());
+                if (protocol.isPortConfigurable() && !Equals(protocol.getDefaultPort(), bookmark.getPort()))
+                {
+                    pathBuilder.Append(":" + bookmark.getPort());
+                }
+            }
+            targetBuilder.Path = pathBuilder.ToString();
+            targetBuilder.Query = "user=" + credentials.getUsername();
```

Comment:
Uses format `application.container.name:protocol.identifier(:port)?\?user=username`.
Should this be changed/updated?

Reply:
Upping this one.

`application.container.name` for Cyberduck is `duck`, for Mountain Duck Windows this is `Cyberduck`. This way Cyberduck and Mountain Duck don't share credentials, which is kind of unfortunate. Which key should be used for these shared elements? Taking into account branded builds.